src/katgpucbf/xbgpu/bsend.py

```diff
@@ -0,0 +1,401 @@
+                    (heaps_per_fengine_per_chunk, self.n_beams),
+                    np.uint32,
+                    context=context,
+                ),
+                channel_offset=channel_offset,
+                timestamp_step=timestamp_step,
+            )
+            self._chunks_queue.put_nowait(chunk)
+            buffers.append(chunk.data)
+
+        # TODO: Follow suit with other Engines to calculate `rate` below
+        stream_config = spead2.send.StreamConfig(
+            max_packet_size=packet_payload + BSend.header_size,
+            # + 1 below for the descriptor per beam
+            max_heaps=(n_tx_items * heaps_per_fengine_per_chunk + 1) * self.n_beams,
+            rate_method=spead2.send.RateMethod.AUTO,
+            rate=0.0,  # TODO: Update to use `send_rate_bytes_per_second`, this sends as fast as possible
+        )
+        self.stream = stream_factory(stream_config, buffers)
+
+        item_group = spead2.send.ItemGroup(flavour=FLAVOUR)
+        item_group.add_item(
+            FREQUENCY_ID,
+            "frequency",  # Misleading name, but it's what the ICD specifies
+            "Value of the first channel in collections stored here.",
+            shape=[],
+            format=IMMEDIATE_FORMAT,
+        )
+        item_group.add_item(
+            TIMESTAMP_ID,
+            "timestamp",
+            "Timestamp provided by the MeerKAT digitisers and scaled to the digitiser sampling rate.",
+            shape=[],
+            format=IMMEDIATE_FORMAT,
+        )
+        item_group.add_item(
+            BF_RAW_ID,
+            "bf_raw",
+            "Beamformer output for frequency-domain beam.",
+            shape=buffers[0].shape[2:],
+            dtype=buffers[0].dtype,
+        )
+
+        self.descriptor_heap = item_group.get_heap(descriptors="all", data="none")
+
+    def enable_substream(self, stream_id: int, enable: bool = True) -> None:
+        """Enable/Disable a substream's data transmission.
+
+        :class:`.BSend` operates as a large single stream with multiple
+        substreams. Each substream is its own data product and is required
+        to be enabled/disabled independently.
+
+        Parameters
+        ----------
+        stream_id
+            Index of the substream's data product.
+        enable
+            Boolean indicating whether the `stream_id` should be enabled or
+            disabled.
+        """
+        self.tx_enabled[stream_id] = enable
+
+    async def get_free_chunk(self) -> Chunk:
+        """Obtain a :class:`.Chunk` for transmission.
+
+        Raises
+        ------
+        asyncio.CancelledError
+            If the chunk's send future is cancelled.
+        """
+        chunk = await self._chunks_queue.get()
+        try:
+            await chunk.future
+        except asyncio.CancelledError:
+            raise asyncio.CancelledError("Sending chunk was cancelled")
+        except Exception:
+            logger.exception("Error sending chunk")
```

Comment:
I'm happy with this logging and otherwise suppressing the error. The point of waiting for the future is just to be sure that we're not going to overwrite data that is still being transmitted. If sending failed, then it's no longer being transmitted and it's safe to return the chunk.

Reply:
I added a small update to the docstring to mention this.